src/nupic/research/temporal_memory.py

```diff
@@ -23,18 +23,124 @@
 Temporal Memory implementation in Python.
 """
 
-from collections import defaultdict, namedtuple
+from collections import defaultdict
 from operator import mul
 
 from nupic.bindings.math import Random
 from nupic.research.connections import Connections
 
-
+from sys import maxint as MAX_INT
+from bisect import bisect_left
 
 EPSILON = 0.000001
 
 
 
+def binSearch(arr, val):
+  i = bisect_left(arr, val)
+  if i != len(arr) and arr[i] == val:
+    return i
+  return -1
+
+
+
+def excitedColumnsGenerator(activeColumns,
+                            activeSegments,
+                            matchingSegments,
+                            cellsPerColumn,
+                            connections):
+  """
+  Generator used for iterating over the lists of active columns,
+  active segments, and matching segments, each sorted by the column they
+  correspond to.
+
+  @param activeColumns    (list)   List of currently active columns
+  @param activeSegments   (list)   List of segments active from lateral input
+  @param matchingSegments (list)   List of segments matching from lateral input
+  @param cellsPerColumn   (int)    Number of cells per column in the tm
+  @param connections      (Object) Connections instance of the tm
+
+  @return (dict){
+                  `column`               (int),
+                  `isActiveColumn`       (bool),
+                  `activeSegmentsBegin`  (int),
+                  `activeSegmentsEnd`    (int),
+                  `matchingSegmentsBegin`(int),
+                  `matchingSegmentsEnd`  (int)
+                }
+
+  Notes:
+    The ints returned (activeSegmentsBegin/End, matchingSegmentsBegin/End)
+    represent bounds of indices into the lists activeSegments and
+    matchingSegments. Together they form an interval of segments for the given
+    column with the schema (note begin is closed and end is open): [begin, end).
+  """
+  activeColumnsProcessed = 0
+  activeSegmentsProcessed = 0
+  matchingSegmentsProcessed = 0
+
+  activeColumnsNum    = len(activeColumns)
```

Comment:
Don't line things up like this. It's super annoying when you make changes later. Some people still do it for doc strings which I guess is personal preference but avoid it for actual code.